test/jdk/java/lang/Float16/BasicFloat16ArithTests.java

```diff
@@ -509,6 +510,56 @@ private static void testSimple() {
             }
         }
 
+        private static void testRounding() {
+            final float ulpOneFp16 = ulp(valueOf(1.0f)).floatValue();
+
+            float [][] testCases = {
+                // The product is equal to
+                // (MAX_VALUE + 1/2 * ulp(MAX_VALUE) + MAX_VALUE = (0x1.ffcp15 + 0x0.002p15)+ 0x1.ffcp15
+                // so overflows.
+                {0x1.3p1f, 0x1.afp15f, -MAX_VAL_FP16,
+                 InfinityF},
+
+                // Product exactly equals 0x1.ffep15, the overflow
+                // threshold; subtracting a non-zero finite value will
+                // result in MAX_VALUE, adding zero or a positive
+                // value will overflow.
+                {0x1.2p10f, 0x1.c7p5f, -0x1.0p-14f,
+                 MAX_VAL_FP16},
+
+                {0x1.2p10f, 0x1.c7p5f, -0.0f,
+                 InfinityF},
+
+                {0x1.2p10f, 0x1.c7p5f, +0.0f,
+                 InfinityF},
+
+                {0x1.2p10f, 0x1.c7p5f, +0x1.0p-14f,
+                 InfinityF},
+
+                {0x1.2p10f, 0x1.c7p5f, InfinityF,
+                 InfinityF},
+
+                // PRECISION bits in the subnormal intermediate product
+                {0x1.ffcp-14f, 0x1.0p-24f, 0x1.0p13f, // Can be held exactly
+                 0x1.0p13f},
+
+                {0x1.ffcp-14f, 0x1.0p-24f, 0x1.0p14f, // *Cannot* be held exactly
+                 0x1.0p14f},
+
+                // Check values where the exact result cannot be
+                // exactly stored in a double.
+                {0x1.0p-24f, 0x1.0p-24f, 0x1.0p10f,
+                 0x1.0p10f},
+
+                {0x1.0p-24f, 0x1.0p-24f, 0x1.0p14f,
+                 0x1.0p14f},
```

Comment:
Your tests only covers overflow conditions, a few going below minimum sub-normal value (0x0.004P-14) and minimum normal value (0x1.000P-14) will also be useful.

Reply:
My query was in following context, citing excerpt from JVM Specification 22 section 2.11.3.

"The Java Virtual Machine does not indicate overflow or underflow during
operations on floating-point data types. That is, floating-point instructions never
cause the Java Virtual Machine to throw a run-time exception (not to be confused
with an IEEE 754 floating-point exception). **An operation that overflows produces
a signed infinity; an operation that underflows produces a subnormal value or a
signed zero;** an operation that has no unique mathematically defined result produces
NaN. All numeric operations with NaN as an operand produce NaN as a result"

Reply:
> My query was in following context, citing excerpt from JVM Specification 22 section 2.11.3.
> 
> "The Java Virtual Machine does not indicate overflow or underflow during operations on floating-point data types. That is, floating-point instructions never cause the Java Virtual Machine to throw a run-time exception (not to be confused with an IEEE 754 floating-point exception). **An operation that overflows produces a signed infinity; an operation that underflows produces a subnormal value or a signed zero;** an operation that has no unique mathematically defined result produces NaN. All numeric operations with NaN as an operand produce NaN as a result"

Sure; that is reasonable.

The correctness of the Float16.fma method relies on the correctness of the double to Float16 conversion. As part of a larger test update, I was planned on expanding the double to Float16 testing. In particular, a test methodology like:

For each finite Float16 value, test the double values around the half-way point to make sure the rounding is correct in those cases. Since there are so few Float16 values, I expect this test to run quickly.